admin/server/github.go

```diff
@@ -86,32 +87,82 @@ func (s *Server) GetGithubUserStatus(ctx context.Context, req *adminv1.GetGithub
 	}
 
+	isUserAppInstalled := false
+	userInstallationPermission := ""
 	installation, _, err := s.admin.Github.AppClient().Apps.FindUserInstallation(ctx, user.GithubUsername)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get user installation: %w", err)
+		if !strings.Contains(err.Error(), "404") {
+			return nil, fmt.Errorf("failed to get user installation: %w", err)
+		}
+	} else {
+		isUserAppInstalled = true
+		userInstallationPermission = "read"
+		// older git app would ask for Contents=read permission whereas new one asks for Contents=write and && Administration=write
+		if installation.Permissions != nil && installation.Permissions.Contents != nil {
+			if installation.Permissions.Administration != nil && strings.EqualFold(*installation.Permissions.Administration, "write") && strings.EqualFold(*installation.Permissions.Contents, "write") {
+				userInstallationPermission = "write"
+			}
+		} else {
+			return nil, status.Errorf(codes.Internal, "failed to get user installation permissions for user %s", user.GithubUsername)
+		}
 	}
 
-	gitClient, err := s.admin.Github.InstallationClient(*installation.ID)
+	client := github.NewTokenClient(ctx, token)
+	// List all the private organizations for the authenticated user
+	orgs, _, err := client.Organizations.List(ctx, "", nil)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get installation client: %w", err)
+		return nil, status.Errorf(codes.Internal, "failed to get user organizations: %s", err.Error())
 	}
-
-	orgs, _, err := gitClient.Organizations.List(ctx, user.GithubUsername, nil)
+	// List all the public organizations for the authenticated user
+	publicOrgs, _, err := client.Organizations.List(ctx, user.GithubUsername, nil)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to get user organizations: %s", err.Error())
 	}
 
+	orgs = append(orgs, publicOrgs...)
+	seen := make(map[string]bool)
+
 	var orgNames []string
+	var orgWithApp []*adminv1.OrganizationWithApp
 	for _, org := range orgs {
+		// dedupe orgs
+		if seen[org.GetLogin()] {
+			continue
+		}
+		seen[org.GetLogin()] = true
+
 		orgNames = append(orgNames, org.GetLogin())
+		i, _, err := s.admin.Github.AppClient().Apps.FindOrganizationInstallation(ctx, org.GetLogin())
+		if err != nil {
+			if strings.Contains(err.Error(), "404") {
+				continue
+			}
+			return nil, status.Errorf(codes.Internal, "failed to get organization installation: %s", err.Error())
+		}
+		permission := "read"
+		// older git app would ask for Contents=read permission whereas new one asks for Contents=write and && Administration=write
+		if i.Permissions != nil && i.Permissions.Contents != nil {
+			if i.Permissions.Administration != nil && strings.EqualFold(*i.Permissions.Administration, "write") && strings.EqualFold(*i.Permissions.Contents, "write") {
+				permission = "write"
+			}
+		} else {
+			return nil, status.Errorf(codes.Internal, "failed to get organization permissions for org %s", org.GetLogin())
+		}
```

Comment:
Can get rid of a level of nesting by reverting the `if` expression and erroring early